src/sbvr-api/permissions.ts

```diff
@@ -0,0 +1,1058 @@
+		if (_.isArray(v)) {
+			rewriteODataOptions(request, v, lambda)
+		} else if (_.isObject(v)) {
+			const propertyName = v.name
+			if (propertyName != null) {
+				if (v.lambda != null) {
+					const newLambda = _.clone(lambda)
+					newLambda[v.lambda.identifier] = sbvrUtils.resolveNavigationResource(request, propertyName)
+					// TODO: This should actually use the top level resource context,
+					// however odata-to-abstract-sql is bugged so we use the lambda context to match that bug for now
+					const subRequest = resolveSubRequest(request, lambda, propertyName, v)
+					rewriteODataOptions(subRequest, v, newLambda)
+				} else if (v.options != null) {
+					_.each(v.options, (option) => {
+						const subRequest = resolveSubRequest(request, lambda, propertyName, v)
+						rewriteODataOptions(subRequest, option, lambda)
+					})
+				} else if (v.property != null) {
+					const subRequest = resolveSubRequest(request, lambda, propertyName, v)
+					rewriteODataOptions(subRequest, v, lambda)
+				} else {
+					rewriteODataOptions(request, v, lambda)
+				}
+			} else {
+				rewriteODataOptions(request, v, lambda)
+			}
+		}
+	})
+}
+
+const memoizedRewriteODataOptions = memoizeWeak(
+		(abstractSqlModel: AbstractSqlModel, vocabulary: string, resourceName: string, _filter: string, tree: ODataQuery) => {
+			tree = _.cloneDeep(tree)
+			rewriteODataOptions({ abstractSqlModel, vocabulary, resourceName }, [tree])
+			return tree
+		}, {
+			normalizer: (_abstractSqlModel, [ vocabulary, resourceName, filter ]) =>
+				filter + vocabulary + resourceName
+		}
+	)
+
+const parseRewrittenPermissions = (abstractSqlModel: AbstractSqlModel, vocabulary: string, resourceName: string, filter: string, odataBinds: ODataBinds) => {
+	let { tree, extraBinds } = $parsePermissions(filter)
+	tree = memoizedRewriteODataOptions(abstractSqlModel, vocabulary, resourceName, filter, tree)
+	return rewriteBinds({ tree, extraBinds }, odataBinds)
+}
+
+const addODataPermissions = (permissionsLookup: PermissionLookup, permissionType: PermissionCheck, vocabulary: string, resourceName: string, odataQuery: ODataQuery, odataBinds: ODataBinds, abstractSqlModel: AbstractSqlModel) => {
+	const conditionalPerms = $checkPermissions(permissionsLookup, permissionType, vocabulary, resourceName)
+
+	if (conditionalPerms === false) {
+		throw new PermissionError()
+	}
+	if (conditionalPerms !== true) {
+		const permissionFilters = nestedCheck(conditionalPerms, (permissionCheck) => {
+			try {
+				// We use an object with filter key to avoid collapsing our filters later.
+				return {
+					filter: parseRewrittenPermissions(abstractSqlModel, vocabulary, resourceName, permissionCheck, odataBinds)
+				}
+			} catch (e) {
+				console.warn('Failed to parse conditional permissions: ', permissionCheck)
+				throw new PermissionParsingError(e)
+			}
+		})
+
+		// if (permissionFilters === false) {
+		// 	throw new PermissionError()
+		// }
+		// if (permissionFilters !== true) {
```

Comment:
Why did you comment this code?

Reply:
Because it's not actually necessary - `permissionFilters` can never be true/false here as you either need to pass in true/false or have the callback to `nestedCheck` return true/false in order for `nestedCheck` to return true/false.  And we know that the `conditionalPerms` we pass in cannot be true/false because we've already checked for it.